x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -113,6 +137,71 @@ public AuthenticationService(
             new ApiKeyAuthenticator(apiKeyService, nodeName),
             new RealmsAuthenticator(numInvalidation, lastSuccessfulAuthCache)
         );
+        this.apiKeyService = apiKeyService;
+    }
+
+    public void authenticateRemoteAccess(
+        final String action,
+        final TransportRequest request,
+        final boolean allowAnonymous,
+        final ActionListener<Authentication> authenticationListener
+    ) {
+        if (false == (threadContext.getHeader(AuthenticationField.AUTHENTICATION_KEY) == null)) {
+            authenticationListener.onFailure(new IllegalArgumentException("authentication header not allowed"));
+            return;
+        }
+
+        final Authenticator.Context context = new Authenticator.Context(
+            threadContext,
+            new AuditableTransportRequest(auditTrailService.get(), failureHandler, threadContext, action, request),
+            null,
+            allowAnonymous,
+            realms
+        );
+        final AuthenticationToken authenticationToken = apiKeyService.getCredentialsFromRemoteAccessHeader(threadContext);
+        context.addAuthenticationToken(authenticationToken);
+        authenticatorChain.authenticateAsync(context, ActionListener.wrap(authentication -> {
+            final RemoteAccessAuthentication remoteAccessAuthentication = RemoteAccessAuthentication.readFromContext(threadContext);
+            final Map<String, String> existingRequestHeaders = threadContext.getRequestHeadersOnly();
+            try (ThreadContext.StoredContext ignored = threadContext.stashContext()) {
+                // drop authentication and remote access authentication headers
+                existingRequestHeaders.forEach((k, v) -> {
+                    if (false == Set.of(
+                        AuthenticationField.AUTHENTICATION_KEY,
+                        SecurityServerTransportInterceptor.REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY,
+                        RemoteAccessAuthentication.REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY
+                    ).contains(k)) {
+                        threadContext.putHeader(k, v);
+                    }
+                });
```

Comment:
I think we should use `ContextPreservingActionListener` here so that original context is restored once authentication finishes with `authenticatorChain` so that we can write the final authentication object, as oppose to nest another threadContext inside which has the risk that the old authentication may be mis-used if downstream code pops the nested threadContext.